backend/optimize.ml

```diff
@@ -0,0 +1,250 @@
+  match get_cells cell 2 with
+  | [fst; snd] -> (
+    let fst_val = DLL.value fst in
+    let snd_val = DLL.value snd in
+    match fst_val.desc with
+    | Op (Move | Spill | Reload) -> (
+      let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+      match snd_val.desc with
+      | Op (Move | Spill | Reload) ->
+        let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+        if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+        then (
+          if !Flambda_backend_flags.cfg_peephole_optimize_track
+          then update_csv "remove_useless_mov";
+          DLL.delete_curr snd;
+          Some (prev_at_most go_back_const fst))
+        else None
+      | _ -> None)
+    | _ -> None)
+  | _ -> None
+
+(** Logical condition for simplifying the following case: 
+  {| 
+    <op 1> const1, r 
+    <op 2> const2, r 
+  |}
+
+  to: 
+  {|
+    <op 1> (const1 <op 2> const2), r 
+  |}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+(* CR-soon gtulba-lecu for gtulba-lecu: implement the rest of Intop_imm cases
+   (only Imulh left to optimize if possible)
+
+   it may be the case that after Iadd, Isub folding the result is 0, treat this
+   case. *)
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1 : Mach.integer_operation), (op2 : Mach.integer_operation) with
+  | Mach.Iand, Mach.Iand -> Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior -> Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor -> Some (Mach.Ixor, imm1 lxor imm2)
+  | Ilsl, Ilsl ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsl, imm1 + imm2)
+    else None
+  | Ilsr, Ilsr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Ilsr, imm1 + imm2)
+    else None
+  | Iasr, Iasr ->
+    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
+    then Some (Mach.Iasr, imm1 + imm2)
+    else None
+  | Iadd, Iadd ->
+    if Misc.no_overflow_add imm1 imm2
+    then Some (Mach.Iadd, imm1 + imm2)
+    else None
+  | Iadd, Isub ->
+    if imm1 >= imm2
+    then Some (Mach.Iadd, imm1 - imm2)
+    else Some (Mach.Isub, imm2 - imm1)
+  | Isub, Isub ->
+    if Misc.no_overflow_add imm1 imm2
+    then Some (Mach.Isub, imm1 + imm2)
+    else None
+  | Isub, Iadd ->
+    if imm1 >= imm2
+    then Some (Mach.Isub, imm1 - imm2)
+    else Some (Mach.Iadd, imm2 - imm1)
+  | Ilsl, Imul ->
+    if Misc.no_overflow_mul (1 lsl imm1) imm2
+    then Some (Mach.Imul, (1 lsl imm1) * imm2)
+    else None
+  | Imul, Ilsl ->
+    if Misc.no_overflow_mul imm1 (1 lsl imm2)
+    then Some (Mach.Imul, imm1 * (1 lsl imm2))
+    else None
+  | Imul, Imul ->
+    if Misc.no_overflow_mul imm1 imm2
+    then Some (Mach.Imul, imm1 * imm2)
+    else None
+  | Idiv, Idiv ->
+    if Misc.no_overflow_mul imm1 imm2
+    then Some (Mach.Idiv, imm1 * imm2)
+    else None
```

Comment:
I think the condition in this one
is not sufficient, for instance if
`imm1` is `-1`.

Reply:
For instance, I think the following function
return the results before/after optimization:
```
let idiv (x : int64) (imm0 : int) (imm1 : int) : int64 * int64 =
  let open Int64 in
  div (div x (of_int imm0)) (of_int imm1),
  div x (of_int (imm0 * imm1));;
```
and `idiv Int64.min_int (-1) max_int` returns:
```
(-2L, 2L)
```

Reply:
This is a very weird edge-case indeed. How this works, if I understand correctly is that:
`Int64.min_int = -2^63 = 10...0_(2)`
`max_int = 2^62 - 1`
`Int64.min_int / -1 = 2^63 = Int64.max_int + 1` which is actually `Int64.min_int` (because it overflows)
So the unoptimized result is actually only `Int64.min_int / max_int`, which indeed is `-2`.

As a sidenote in C++ the following piece of code prints the same value 3 times:
```
#include <cstdio>
#include <climits>
#include <algorithm>

int main() {
    printf("%lld\n", abs(LLONG_MIN));
    printf("%lld\n", LLONG_MIN);
    printf("%lld\n", LLONG_MIN * -1);
    return 0;
}
```

I think we should do the following:
1. add a check to just ignore this optmization (and the one for Imul) if one of the arguments is min_int
2. do the check for overflow for the absolute values of imm1 and imm2

As another sidenote, I am pretty sure that the division is optimized somewhere upstream in the compiler since to a multiplication since I could not find a Idiv in the cfg IR when I checked for it. (Nevertheless, I might be wrong about this)